### PR TITLE
fix(s3): double free of path when S3Client operation throws

### PR DESCRIPTION
**Pull request by robobun**
## What does this PR do?

Fixes an ASAN use-after-poison (double free) in `S3Client` methods when an operation throws after the internal blob has been constructed.

### Root cause

In methods like `S3Client.prototype.presign`:

```zig
const path = try jsc.Node.PathLike.fromJS(globalThis, &args) orelse { ... };
errdefer path.deinit();
var blob = try S3File.constructS3FileWithS3CredentialsAndOptions(globalThis, path, ...);
defer blob.detach();
return S3File.getPresignUrlFrom(&blob, globalThis, options);
```

`constructS3FileWithS3CredentialsAndOptions` stores `path` directly in the blob's store (no copy). If `getPresignUrlFrom` then throws (missing credentials, `expiresIn: -1`, a throwing option getter, etc.), `defer blob.detach()` frees the path via the store **and** `errdefer path.deinit()` frees it again.

This only manifests visibly when the `PathLike` is an allocated `.encoded_slice`, which happens for paths containing non-ASCII characters (UTF-16 → UTF-8 conversion allocates).

The same pattern existed across `file`/`presign`/`exists`/`size`/`stat`/`write`/`unlink` on both the instance and the static `S3Client` / `S3File` code paths.

### Fix

- `constructS3FileWithS3Credentials` and `constructS3FileWithS3CredentialsAndOptions` now **always** take ownership of `path`: freed on error, stored in the returned blob on success.
- All `S3Client` instance methods drop the `errdefer path.deinit()` since ownership is unconditionally transferred. `write` explicitly frees `path` in the missing-data case before the constructor is reached.
- The `S3File` static methods clear `path_or_blob` before handing the path to the constructor so their `errdefer` becomes a no-op once ownership is gone.

### Repro

```js
new Bun.S3Client().presign("bucket/key-ü.txt", { expiresIn: -1 });
```

Before: `AddressSanitizer: use-after-poison` in `PathLike.deinit`.
After: throws `expiresIn must be greather than 0` cleanly.

## How did you verify your code works?

- Reproduced the ASAN crash with both instance and static `presign` using a non-ASCII path.
- Added `test/js/bun/s3/s3-path-double-free.test.ts` covering both the before-blob and after-blob error paths for instance and static methods.
- `bun bd test test/js/bun/s3/s3-path-double-free.test.ts` passes; the same test crashes under the previous ASAN build.
- Original fuzzer crash script runs to completion with exit 0.

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 4:27 PM PT - Apr 23rd, 2026</sup></div>

@robobun, your commit 99bb68d9565dbc377d4d9c3bd06460936d35b77e is building: [`#47624`](https://buildkite.com/bun/bun/builds/47624)

File: test/js/bun/s3/s3-path-double-free.test.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  test("static file() throwing before blob creation", () => {
+    expect(() =>
+      Bun.S3Client.file(nonAsciiPath, {
+        get type() {
+          throw new Error("boom");
+        },
+      }),
+    ).toThrow("boom");
+  });
+});
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/s3/s3-path-double-free.test.ts:110]: Added coverage for `exists`, `size`, `stat`, and `unlink` (instance + static) in de5bcd3.

File: test/js/bun/s3/s3-path-double-free.test.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      const client = new Bun.S3Client();
+      expect(() =>
+        client[method](nonAsciiPath, {
+          get type() {
+            throw new Error("boom");
+          },
+        }),
+      ).toThrow("boom");
+    },
+  );
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/s3/s3-path-double-free.test.ts]: `test.each` is used more often than `describe.each` in `test/js/bun/` (23 vs 15 occurrences). For a single assertion per parameter value, `test.each` is the more direct form; wrapping each in a `describe` block with one inner `test` adds nesting without adding clarity. Leaving as-is.

**Comment by robobun** [comment]
CI on build 47626: all 60 jobs that ran passed, including `debian-13-x64-asan-test-bun` (the ASAN shard that exercises this fix). The two red checks are `darwin-{13,14}-aarch64-test-bun` with status **Expired** — they timed out waiting for a buildkite agent and never started. Not related to this change.

File: src/runtime/webcore/S3Client.zig [hunk 3]
```diff
@@ -251,7 +245,6 @@ pub const S3Client = struct {
         const path: jsc.Node.PathLike = try jsc.Node.PathLike.fromJS(globalThis, &args) orelse {
             return globalThis.ERR(.MISSING_ARGS, "Expected a path to unlink", .{}).throw();
         };
-        errdefer path.deinit();
         const options = args.nextEat();
         var blob = try S3File.constructS3FileWithS3CredentialsAndOptions(globalThis, path, options, ptr.credentials, ptr.options, ptr.acl, ptr.storage_class, ptr.request_payer);
```
Comment (Jarred-Sumner) [inline src/runtime/webcore/S3Client.zig:249]: Doesn't this create a memory leak right here if it throws?

Comment (robobun) [reply src/runtime/webcore/S3Client.zig:249]: No — `constructS3FileWithS3CredentialsAndOptions` now unconditionally takes ownership of `path` (see the other hunk in `S3File.zig`). Inside it:

```zig
var aws_options = S3.S3Credentials.getCredentialsWithOptions(...) catch |err| {
    path.deinit();
    return err;
};
defer aws_options.deinit();

const store = brk: { ... Blob.Store.initS3(path, ...) ... }; // store owns path from here
errdefer store.deinit();                                      // frees path on any later error
```

So every error path inside the constructor frees `path` exactly once, and on success the returned blob owns it (freed by `defer blob.detach()`).

Between `fromJS` and the constructor call there is only `args.nextEat()`, which returns `?jsc.JSValue` and cannot throw.

Keeping the old `errdefer path.deinit()` here is what caused the double free: when `s3.unlink(blob.store.?, globalThis, options)` (or `getPresignUrlFrom` in `presign`) threw after the blob existed, both `defer blob.detach()` (store → path) and `errdefer path.deinit()` fired on the same allocation.
